Python/cheritrace.i

```diff
@@ -1,80 +1,250 @@
+	  {
+	       code = PyObject_GetAttrString(callable, "__code__");
+	       if (!code) {
+		    PyErr_Clear();
+		    /* may be a callable class */
+		    run = PyObject_GetAttrString(callable, "__call__");
+		    if (!run)
+			 return 0;
+		    func = PyObject_GetAttrString(run, "__func__");
+		    Py_DECREF(run);
+		    if (!func)
+			 return 0;
+		    code = PyObject_GetAttrString(func, "__code__");
+		    Py_DECREF(func);
+		    if (!code)
+			 return 0;
+		    c_argcount = -1; /* account for self */
+	       }
+	       argcount = PyObject_GetAttrString(code, "co_argcount");
+	       if (!argcount)
+	       {
+		    Py_DECREF(code);
+		    return 0;
+	       }
+	       c_argcount += PyInt_AsLong(argcount);
+	       Py_DECREF(argcount);
+	       return c_argcount;
+	  }
+	  return 0;
+     }
+}
+
+%typemap(typecheck,
+	 precedence=SWIG_TYPECHECK_INTEGER,
+	 fragment="PyCheritrace_GetArgcount")
+cheri::streamtrace::trace::scanner
+{
+     $1 = PyCheritrace_GetArgcount($input) == 2;
+}
+
+%typemap(typecheck,
+	 precedence=SWIG_TYPECHECK_INTEGER,
+	 fragment="PyCheritrace_GetArgcount")
+cheri::streamtrace::trace::detailed_scanner
+{
+     $1 = PyCheritrace_GetArgcount($input) == 3;
+}
+
+%typemap(typecheck,
+	 precedence=SWIG_TYPECHECK_INTEGER,
+	 fragment="PyCheritrace_GetArgcount")
+cheri::streamtrace::trace::filter_predicate
+{
+     $1 = PyCheritrace_GetArgcount($input) == 1;
+}
 
-//%nodefaultctor cheri::streamtrace::Scanner;
 %include "../streamtrace.hh";
 %include "../disassembler.hh";
-%include "callback.hh"
-
-%extend cheri::streamtrace::trace {
-
-    void scan_trace(cheri::streamtrace::Scanner* scn) {
-      auto bound_cbk = std::bind(&cheri::streamtrace::Scanner::c_scanner,
-				 scn,
-				 std::placeholders::_1,
-				 std::placeholders::_2);
-      $self->scan(cheri::streamtrace::trace::scanner(bound_cbk));
-    }
-
-    void scan_trace(cheri::streamtrace::Scanner* scn, uint64_t start, uint64_t end, int opts=0) {
-      auto bound_cbk = std::bind(&cheri::streamtrace::Scanner::c_scanner,
-				 scn,
-				 std::placeholders::_1,
-				 std::placeholders::_2);
-      $self->scan(cheri::streamtrace::trace::scanner(bound_cbk), start, end, opts);
-    }
-
-    void scan_trace(cheri::streamtrace::DetailedScanner* scn, uint64_t start, uint64_t end, int opts=0) {
-      auto bound_cbk = std::bind(&cheri::streamtrace::DetailedScanner::c_scanner,
-				 scn,
-				 std::placeholders::_1,
-				 std::placeholders::_2,
-				 std::placeholders::_3);
-      $self->scan(cheri::streamtrace::trace::detailed_scanner(bound_cbk), start, end, opts);
-    }
-
-    std::shared_ptr<trace_view> filter_trace(cheri::streamtrace::Filter* filter) {
-      auto bound_cbk = std::bind(&cheri::streamtrace::Filter::c_filter,
-				 filter,
-				 std::placeholders::_1);
-      return $self->filter(cheri::streamtrace::trace::filter_predicate(bound_cbk));
-    }
+
+%extend cheri::streamtrace::debug_trace_entry
+{
+     /* Workaround missing support of nested unions */
```

Comment:
You can make these methods visible in C++ if you like - they're probably better style anyway.